mkfs/rootdir.c

```diff
@@ -286,33 +397,108 @@ static int write_extent(struct btrfs_trans_handle *trans,
 		bytes_read += ret_read;
 	}
 
-	memset(buf + to_read, 0, padding);
+	if (bytes_read <= sectorsize)
+		do_comp = false;
+
+#if COMPRESSION_ZSTD
+	if (do_comp) {
+		comp_ret = compress_extent(btrfs_inode, sectorsize, buf,
+					   bytes_read, comp_buf);
+		if (comp_ret < 0)
+			return comp_ret;
+		else if (comp_ret == 0)
+			do_comp = false;
+
+		/* If we've just marked this inode nocompress, increase the
+		 * buffer size and read the rest of the extent.  */
+		if (btrfs_stack_inode_flags(btrfs_inode) & BTRFS_INODE_NOCOMPRESS) {
+			buf_size = SZ_1M;
+
+			if (*left < buf_size) {
+				to_read = *left;
+
+				if (to_read % sectorsize)
+					padding = sectorsize - (to_read % sectorsize);
+				else
+					padding = 0;
+			} else {
+				to_read = buf_size;
+				padding = 0;
+			}
+
+			while (bytes_read < to_read) {
+				ssize_t ret_read;
+
+				ret_read = pread(fd, buf + bytes_read,
+						 to_read - bytes_read,
+						 *file_pos + bytes_read);
+				if (ret_read == -1) {
+					error("cannot read %s at offset %llu length %llu: %m",
+					      path_name, *file_pos + bytes_read,
+					      to_read - bytes_read);
+					return -errno;
+				}
+
+				bytes_read += ret_read;
+			}
+		}
+	}
 
-	ret = btrfs_reserve_extent(trans, root, to_read + padding, 0, 0,
+	if (do_comp) {
+		u64 flags, features;
+
+		to_write = DIV_ROUND_UP(comp_ret, sectorsize) * sectorsize;
+		write_buf = comp_buf;
+		memset(write_buf + comp_ret, 0, to_write - comp_ret);
```

Comment:
I think this patter can also applied to the previous patch to do the padding, without introducing a dedicated `padding` variable.